### PR TITLE
Support ESM and .mts/.cts extensions with jiti loader

### PR DESCRIPTION
**Pull request by beerose**
## Description

A second approach to adding ESM support. [The previous one](https://github.com/kamilkisiela/graphql-config/pull/1281) included using `tsc` and custom caching and this caching bit had a bunch of edge cases that I wasn't aware of in the beginning. 

I replaced it with `jiti` loader and added more test cases. `jiti` is much faster than running `tsc` (similar time to cosmic loader).

Related to issues in graphql-code-generator: https://github.com/dotansimha/graphql-code-generator/issues/8509 and https://github.com/dotansimha/graphql-code-generator/issues/8862. More context in this PR: https://github.com/dotansimha/graphql-code-generator/pull/9086.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Extended existing tests.

**Test Environment**:

- OS:
- GraphQL Config Version:
- NodeJS:

## Checklist:

- [ ] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests and linter rules pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by B2o5T — APPROVED** [review]

File: src/helpers/cosmiconfig.ts [hunk 1]
```diff
@@ -40,9 +40,24 @@ export function createCosmiConfigSync(moduleName: string, legacy: boolean) {
 }
 
 const loadTypeScript: Loader = (...args) => {
+  try {
+    // eslint-disable-next-line @typescript-eslint/no-var-requires
+    const { TypeScriptLoader } = require('cosmiconfig-typescript-loader');
+    return TypeScriptLoader({ transpileOnly: true })(...args);
```
Comment (B2o5T) [inline src/helpers/cosmiconfig.ts:46]: @beerose do we still need `cosmiconfig-typescript-loader`?

File: src/helpers/cosmiconfig.ts [hunk 2]
```diff
@@ -40,9 +40,24 @@ export function createCosmiConfigSync(moduleName: string, legacy: boolean) {
...
+    }
+    throw err;
+  }
+};
+
+const jitiLoader: Loader = (filepath) => {
   // eslint-disable-next-line @typescript-eslint/no-var-requires
-  const { TypeScriptLoader } = require('cosmiconfig-typescript-loader');
-  return TypeScriptLoader({ transpileOnly: true })(...args);
+  const jiti = require('jiti')(__filename, {
```
Comment (B2o5T) [inline src/helpers/cosmiconfig.ts:57]: I guess you can use `import` statement instead (since this package in `"dependencies"`), `cosmiconfig-typescript-loader` used with `require` because it's a peerDep with peerDependecyMeta

File: src/helpers/cosmiconfig.ts (hunk 1 above)
Comment (beerose) [reply src/helpers/cosmiconfig.ts]: Yeah, it looks like we don't! Just pushed the changes
